src/juju_doctor/probes.py

```diff
@@ -158,3 +171,74 @@ def run(self, artifacts: Artifacts) -> List[ProbeResults]:
             else:
                 results.append(ProbeResults(probe_name=f"{self.name}/{func_name}", passed=True))
         return results
+
+
+def _read_file(filename: Path) -> Optional[Dict]:
+    """Read a file into a string."""
+    try:
+        with open(str(filename), "r") as f:
+            return yaml.safe_load(f)
+    except FileNotFoundError:
+        log.warning(f"Error: File '{filename}' not found.")
+    except yaml.YAMLError as e:
+        log.warning(f"Error: Failed to parse YAML in '{filename}': {e}")
+    except Exception as e:
+        log.warning(f"Unexpected error while reading '{filename}': {e}")
+    return None
+
+
+class RuleSet:
+    """Represents a set of probes defined in a ruleset configuration file.
+
+    Supports recursive aggregation of probes, handling scriptlets and nested rulesets.
+    """
+
+    def __init__(self, probe: Probe, name: Optional[str] = None):
+        """Initialize a RuleSet instance.
+
+        Args:
+            probe (Probe): The Probe representing the ruleset configuration file.
+            name (str): The name of the ruleset.
+        """
+        self.probe = probe
+        self.name = name or self.probe.name
+
+    def aggregate_probes(self) -> List[Probe]:
+        """Obtain all the probes from the RuleSet.
+
+        This method is recursive when it finds another RuleSet probe and returns
+        a list of probes that were found after traversing all the probes in the ruleset.
+        """
+        content = _read_file(self.probe.path)
+        if not content:
+            return []
```

Comment:
For a separate issue: how do we best surface that we failed reading a file?
A warning seems insufficient.